Mathlib/Order/RelSeries.lean

```diff
@@ -276,6 +319,150 @@ def insertNth (p : RelSeries r) (i : Fin p.length) (a : α)
           simp only [Fin.pred_succ, eq_rec_constant]
           congr; ext; exact hm.symm
 
+/--
+A relation series `a₀ -r→ a₁ -r→ ... -r→ aₙ` of `r` gives a relation series of the reverse of `r`
+by reversing the series `aₙ ←r- aₙ₋₁ ←r- ... ←r- a₁ ←r- a₀`.
+-/
+@[simps length]
+def reverse (p : RelSeries r) : RelSeries (fun (a b : α) ↦ r b a) where
+  length := p.length
+  toFun := p ∘ Fin.rev
+  step i := by
+    rw [Function.comp_apply, Function.comp_apply]
+    have hi : i.1 + 1 ≤ p.length
+    · linarith [i.2]
+    convert p.step ⟨p.length - (i.1 + 1), Nat.sub_lt_self (by linarith) hi⟩
+    · ext; simp
+    · ext
+      simp only [Fin.val_rev, Fin.coe_castSucc, Nat.succ_sub_succ_eq_sub, Fin.val_succ]
+      rw [Nat.sub_eq_iff_eq_add, add_assoc, add_comm 1 i.1, Nat.sub_add_cancel] <;>
+      aesop
+
+@[simp] lemma reverse_apply (p : RelSeries r) (i : Fin (p.length + 1)) :
+    p.reverse i = p i.rev := rfl
+
+/--
+given a series `a₀ -r→ a₁ -r→ ... -r→ aₙ` and an `a` such that `a₀ -r→ a` holds, there is
+a series of length `n+1`: `a -r→ a₀ -r→ a₁ -r→ ... -r→ aₙ`.
+-/
+@[simps! length]
+def cons (p : RelSeries r) (newHead : α) (rel : r newHead p.head) : RelSeries r :=
+  (singleton r newHead).append p rel
+
+@[simp] lemma cons_head (p : RelSeries r) (newHead : α) (rel : r newHead p.head) :
+    (p.cons newHead rel).head = newHead := rfl
+
+@[simp] lemma cons_last (p : RelSeries r) (newHead : α) (rel : r newHead p.head) :
+    (p.cons newHead rel).last = p.last := by
+  delta cons
+  rw [append_last]
+
+/--
+If a series `a₀ -r→ a₁ -r→ ...` has positive length, then `a₁ -r→ ...` is another series
+-/
+@[simps]
+def tail (p : RelSeries r) (len_pos : p.length ≠ 0) : RelSeries r where
+  length := p.length - 1
+  toFun := Fin.tail p ∘ (Fin.cast <| Nat.succ_pred_eq_of_pos <| Nat.pos_of_ne_zero len_pos)
+  step i := p.step ⟨i.1 + 1, Nat.lt_pred_iff.mp i.2⟩
+
+/--
+If a series ``a₀ -r→ a₁ -r→ ... -r→ aₙ``, then `a₀ -r→ a₁ -r→ ... -r→ aₙ₋₁` is
+another series -/
+@[simps]
+def eraseLast (p : RelSeries r) : RelSeries r where
+  length := p.length - 1
+  toFun i := p ⟨i, lt_of_lt_of_le i.2 (Nat.succ_le_succ tsub_le_self)⟩
+  step i := p.step ⟨i, lt_of_lt_of_le i.2 tsub_le_self⟩
+
+/--
+Given two series of the form `a₀ -r→ ... -r→ X` and `X -r→ b ---> ...`,
+then `a₀ -r→ ... -r→ X -r→ b ...` is another series obtained by combining the given two.
+
+-/
+@[simps]
+def combine (p q : RelSeries r) (connect : p.last = q.head) : RelSeries r where
+  length := p.length + q.length
+  toFun i :=
+    if H : i.1 < p.length
+    then p ⟨i.1, H.trans (lt_add_one _)⟩
+    else q ⟨i.1 - p.length,
+      Nat.sub_lt_left_of_lt_add (by rwa [not_lt] at H) (by rw [← add_assoc]; exact i.2)⟩
```

Comment:
This could be implemented in terms of `eraseLast` and `append`, but I think this way makes it easier to prove later lemmas like `combine_castAdd` etc.